From 9319c91559e1dfbfc098f85fb733d5eae444c59c Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Thu, 3 Mar 2022 08:22:30 +0000
Subject: [PATCH] xen/arm: Set domain default PARTID and PMG

System register MPAM0_EL1 and MPAM1_EL1 hold information (PARTID and
PMG) to generate MPAM labels for memory requests at EL0 and EL1. This
register is context-switched when switching vCPUs.

Since if Xen needs to partition the resource of a specific VM, it
needs to limit the VM both in EL1 and EL0. This commit sets the default
PARTID and PMG to the MPAM0_EL1 and MPAM1_EL1 registers of each vCPU at
the domain creation stage. The VMID of the domain is used as the PARTID.
As PMG is not in scope for now, this commit sets all default PMGs to 0.
Set the MPAM0_EL1 and MPAM1_EL1 for the domain when the vCPUs of this
domain is created by a helper mpam_set_vcpu_partid_pmg().

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3395
---
 xen/arch/arm/domain.c           |  6 ++++++
 xen/arch/arm/include/asm/mpam.h |  8 ++++++++
 xen/arch/arm/mpam.c             | 15 +++++++++++++++
 3 files changed, 29 insertions(+)

diff --git a/xen/arch/arm/domain.c b/xen/arch/arm/domain.c
index 6bcbdc2c5dc4..cb619121fc90 100644
--- a/xen/arch/arm/domain.c
+++ b/xen/arch/arm/domain.c
@@ -593,6 +593,11 @@ int arch_vcpu_create(struct vcpu *v)
     if ( get_ssbd_state() == ARM_SSBD_RUNTIME )
         v->arch.cpu_info->flags |= CPUINFO_WORKAROUND_2_FLAG;
 
+#ifdef CONFIG_MPAM
+    mpam_set_vcpu_partid_pmg(v, v->domain->arch.mpam.partid,
+                             v->domain->arch.mpam.partid, 0, 0);
+#endif
+
     return rc;
 
 fail:
@@ -801,6 +806,7 @@ int arch_domain_create(struct domain *d,
 #endif
 
 #ifdef CONFIG_MPAM
+    d->arch.mpam.partid = d->arch.p2m.vmid;
     d->arch.mpam.cfg.l3_cpbm = config->arch.mpam.l3_cpbm;
     if ( mpam_enabled )
         mpam_update_domain(d);
diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index 78d1f5d7c6c2..e902503e07ad 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -258,6 +258,11 @@
 #define MSMON_CAPT_EVNT_NOW    BIT(0, UL)
 
 /* Field descriptions for MPAM system registers: */
+#define MPAM_SYSREG_PMG_D      GENMASK(47, 40)
+#define MPAM_SYSREG_PMG_I      GENMASK(39, 32)
+#define MPAM_SYSREG_PARTID_D   GENMASK(31, 16)
+#define MPAM_SYSREG_PARTID_I   GENMASK(15, 0)
+
 #define MPAMIDR_PMG_MAX        GENMASK(40, 32)
 #define MPAMIDR_PARTID_MAX     GENMASK(15, 0)
 
@@ -417,6 +422,9 @@ extern void mpam_restore_state(struct vcpu *n);
 extern void mpam_save_state(struct vcpu *n);
 extern void mpam_sanitise_config(struct xen_arch_domctl_mpam *cfg,
                                  bool domain_create);
+extern void mpam_set_vcpu_partid_pmg(struct vcpu *vcpu,
+                                     uint16_t partid_d, uint16_t partid_i,
+                                     uint8_t pmg_d, uint8_t pmg_i);
 extern void mpam_update_domain(struct domain *d);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
 
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 7d466c2f1fe4..ca499f398b81 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -1474,6 +1474,21 @@ void mpam_restore_state(struct vcpu *n)
     WRITE_SYSREG(n->arch.mpam1_el1, MPAM1_EL1);
 }
 
+void mpam_set_vcpu_partid_pmg(struct vcpu *vcpu,
+                              uint16_t partid_d, uint16_t partid_i,
+                              uint8_t pmg_d, uint8_t pmg_i)
+{
+    uint64_t regval;
+
+    regval = FIELD_PREP(MPAM_SYSREG_PARTID_D, partid_d);
+    regval |= FIELD_PREP(MPAM_SYSREG_PARTID_I, partid_i);
+    regval |= FIELD_PREP(MPAM_SYSREG_PMG_D, pmg_d);
+    regval |= FIELD_PREP(MPAM_SYSREG_PMG_I, pmg_i);
+
+    vcpu->arch.mpam0_el1 = regval;
+    vcpu->arch.mpam1_el1 = regval;
+}
+
 /*
  * Local variables:
  * mode: C
-- 
2.34.1

